### ISSUE TITLE
Built-in additive colour modulation

### ISSUE THREAD

**Issue by snoopdouglas** [godotengine/godot-proposals#1972]
### Describe the project you are working on

Platformer (n/a really)

### Describe the problem or limitation you are having in your project

Additional to colour modulation (which Godot currently implements *subtractively* on `Node2D.modulate`), I often find myself wanting to *additively* modulate a node's colour - that is, to 'brighten' it.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I'd like to have it as a built-in feature, eg. `Node2D.modulate_mode`.

While this is possible through shaders (which are currently how I do this), I find myself needing to brighten my nodes quite often, so would prefer not to have to use shaders.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

To demonstrate, here's a shader for 2D textures that implements an additive modulation (with linear interpolation, similar to how Godot does it subtractively):

```glsl
shader_type canvas_item;
render_mode unshaded;

uniform vec4 modulate = vec4(0, 0, 0, 0);

void fragment() {
	COLOR = mix(texture(TEXTURE, UV), vec4(1, 1, 1, 1), modulate);
}
```

I also find being able to either add *or* subtract useful (using the same modulation colour for both). I personally call this 'duplex' - but I'm unsure whether there's a better word for it.

This can be done by modulating with a colour that defaults to mid-grey, where values < 0.5 subtract, and values > 0.5 add:

```glsl
shader_type canvas_item;
render_mode unshaded;

uniform vec4 modulate = vec4(0.5, 0.5, 0.5, 0.5);

void fragment() {
	vec4 modulate2 = (modulate - 0.5) * 2.0;
	COLOR = mix(texture(TEXTURE, UV), floor(modulate2), abs(modulate2));
}
```

For subtractive, additive and duplex modulation respectively, `Node2D.modulate_mode` could be an enum with `SUBTRACT, ADD, DUPLEX`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

If I won't, yes (as demonstrated above). I think it would though. ;)

### Is there a reason why this should be core and not an add-on in the asset library?

I personally find myself using this a lot. If it's just me though, yes, a plugin would probably be achievable.

**Comment by Calinou** [comment]
You can add a Sprite with a CanvasItemMaterial in Add mode on top of your main sprite top achieve this.

Also, remember that you can set color channels in the Modulate property above 1 to brighten a sprite.

**Comment by snoopdouglas** [comment]
> You can add a Sprite with a CanvasItemMaterial in Add mode on top of your main sprite to achieve this.

Yes, though I personally wouldn't do it this way.

> you can set color channels in the Modulate property above 1 to brighten a sprite

This doesn't work as I'd expect. Firstly, it's not a linear interpolation, and secondly, I'd expect the cyan box at the top (which is being modulated here) to be totally white - but it isn't:

![Screenshot from 2020-12-11 19-01-25](https://user-images.githubusercontent.com/4892574/101943647-60a56880-3be3-11eb-8a18-1024c5d56aab.png)

**Comment by Calinou** [comment]
> This doesn't work as I'd expect. Firstly, it's not a linear interpolation, and secondly, I'd expect the cyan box in the top left (which is being modulated here) to be totally white - but it isn't:

It's not completely white because the color representation in the ColorPicker is just a simulated preview. `Color(100, 100, 100)` will not make fully black pixels appear as white.

**Comment by snoopdouglas** [comment]
> multiplying

Yep, that's exactly the problem :p you can't do additive modulation if there isn't an addition somewhere in the equation.

**Comment by Calinou** [comment]
See also https://github.com/godotengine/godot-proposals/issues/96.

**Comment by merachefet** [comment]
I would find this really useful. I have an example use-case here of what's possible with additive vs. subtractive modulation. On the left, I'm using the existing modulation, and on the right I'm adding color. Using lighter and darker base sprites gives me the widest possible range for each.

As you can see, the effect is very different. Subtraction is preserving the shadow contrast, while addition is preserving the highlights. This gives me the glossy, specular look that I want for this slime monster. I can't get the effect in the lower-right using subtraction.

It's the same as PBR shading, you need to start with a darker color to make something look reflective.

![slime_example](https://github.com/godotengine/godot-proposals/assets/2681169/fa1fd659-aa70-4035-a830-d8bd90681c6d)
